src/Adapters/Redis/RedisStorage.php

```diff
@@ -0,0 +1,99 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Lamoda\Metric\Adapters\Redis;
+
+use Lamoda\Metric\Common\MetricSourceInterface;
+use Lamoda\Metric\Storage\MetricStorageInterface;
+use Lamoda\Metric\Storage\MutableMetricInterface;
+
+final class RedisStorage implements \IteratorAggregate, MetricStorageInterface, RedisStorageInterface
+{
+    /** @var RedisConnectionInterface */
+    private $redisConnection;
+
+    public function __construct(RedisConnectionInterface $redisConnection)
+    {
+        $this->redisConnection = $redisConnection;
+    }
+
+    public function getIterator(): \Traversable
+    {
+        return $this->getMetrics();
+    }
+
+    public function receive(MetricSourceInterface $source): void
+    {
+        $metrics = [];
+        foreach ($source->getMetrics() as $metric) {
+            $metrics[] = new MetricDto(
+                $metric->getName(),
+                $metric->resolve(),
+                $this->convertTagsForStorage($metric->getTags())
+            );
+        }
+        $this->redisConnection->setMetrics($metrics);
+    }
+
+    public function getMetrics(): \Traversable
+    {
+        $metricsData = $this->redisConnection->getAllMetrics();
+        foreach ($metricsData as $metricDto) {
+            yield $this->createMetricEntity(
+                $metricDto->name,
+                $metricDto->value,
+                $this->convertTagsFromStorage($metricDto->tags)
+            );
+        }
+    }
+
+    public function findMetric(string $name, array $tags = []): ?MutableMetricInterface
+    {
+        $value = $this->redisConnection->getMetricValue($name, $this->convertTagsForStorage($tags));
+
+        return $this->createMetricEntity($name, $value, $tags);
+    }
+
+    public function createMetric(string $name, float $value, array $tags = []): MutableMetricInterface
+    {
+        $tags = $this->normalizeTags($tags);
+        $this->setMetricValue($name, $value, $tags);
+
+        return $this->createMetricEntity($name, $value, $tags);
+    }
+
+    public function setMetricValue(string $name, float $value, array $tags = []): void
+    {
+        $this->redisConnection->setMetric($name, $value, $this->convertTagsForStorage($tags));
+    }
+
+    public function adjustMetricValue(string $name, float $value, array $tags = []): void
+    {
+        $this->redisConnection->adjustMetric($name, $value, $this->convertTagsForStorage($tags));
+    }
+
+    private function createMetricEntity(string $name, float $value, array $tags): MutableMetricInterface
+    {
+        return new Metric($name, $value, $tags, $this);
+    }
+
+    private function convertTagsForStorage(array $tags): string
+    {
+        $tags = $this->normalizeTags($tags);
+
+        return json_encode($this->normalizeTags($tags));
+    }
+
+    private function convertTagsFromStorage(string $tags): array
+    {
+        return json_decode($tags, true);
+    }
+
+    private function normalizeTags(array $tags): array
```

Comment:
> This logic can be kept inside Metric class itself.

In this case we will not normalize tags in `RedisStorage::findMetric(string $name, array $tags = [])` also. Is that acceptable?

Reply:
Yeah, in that case I suggest to move everything related to tag normalization into connection, since calling `Connection::getMetricValue` does no tag conversion also (which is error prone)

Reply:
Normalizing tags is moved to `Connection`